enhancements/testing/improved-platform-tests.md

```diff
@@ -0,0 +1,412 @@
+---
+title: improved-platform-tests
+authors:
+  - "@soltysh"
+reviewers:
+  - "@ardaguclu"
+  - "@bertinatto"
+  - "@deads2k"
+  - "@ingvagabund"
+approvers:
+  - "@deads2k"
+api-approvers:
+  - None
+creation-date: 2022-11-16
+last-updated: 2023-05-31
+tracking-link:
+  - https://issues.redhat.com/browse/WRKLDS-720
+replaces:
+  - "enhancements/testing/extended-platform-tests.md"
+---
+
+# Improved Platform Tests
+
+## Summary
+
+Allow maintaining and building end-to-end (e2e) tests outside of
+[origin](https://github.com/openshift/origin/) repository, and running them under
+single `openshift-tests` umbrella, by extracting the necessary test binaries
+during runtime.
+
+## Motivation
+
+The reasoning behind this change are two-folded:
+
+1. Short term goal is to speed up the [kubernetes updates](#kubernetes-updates).
+2. Long term goal is separating all e2e tests so that they can live in their
+   [respective component repos](#tests-co-located-with-code).
+
+### User Stories
+
+- As an OpenShift developer, I want to be able to seamlessly run OpenShift
+  and kubernetes-specific tests using `openshift-tests run`, so that my current
+  workflows continue working as before.
+- As an OpenShift developer, I want to be able to run newest kubernetes tests
+  without the necessity to bump vendored code, to limit the amount of re-vendoring
+  work and ensure all proper test coverage.
+- As an OpenShift developer, I want to be able to extend and monitor cluster
+  during tests execution, to ensure high availability of the platform.
+- As an OpenShift partner or user, I want to be able to verify that my cluster is providing
+  the desired functionality (CSI for example) running one of the provided test suites.
+- As an OpenShift developer, I want to be able to deliver new e2e tests in the
+  same pull request (PR) that introduces the feature being tested, so that I can
+  ensure my new feature code is properly tested prior to merging.
+
+### Goals
+
+1. Reduce the risk updating kubernetes caused by excessive co-located code,
+   and minimize time when we start running newest kubernetes tests.
+2. Remove the requirement that all test code lives in a single repository.
+3. Maintain `openshift-tests` API for running tests.
+4. Ensure that kubernetes tests are still vendored in `openshift-tests` binary,
+   to allow for a fallback when the release payload image is not reachable.
```

Comment:
I see this one as temporary....once we start having all teams moving their e2e tests into their own repos, we're just going to have to accept that the test images need to be available or the test job fails.

and we're not going to want to maintain this vendoring indefinitely just to give us a fallback (not to mention that falling back is risky since now you're not necessarily running the tests you thought you were)

Reply:
For some cases the fallback will be a must, microshift comes to my mind immediately, since they don't provide release image. I'm not sure about hypershift, either. So at least for k8s-related tests I think that might be longer than just temporary. We'll need to re-assess this at a later point in time. That's why I've explicitly called out k8s tests only for backwards compatibility.

Reply:
I've added this to "Even more future work" section, there's still work about upgrade tests, so this is a good future thing to look at .

Reply:
if nothing else, fallback needs to be an opt-in behavior (do not fallback to another version of the test code silently or by default).

but this is another reason why producing a single image that contains all the test binaries would be a better solution in my mind.

cc @dhellmann since he can better speak to what microshift does today or would want to support in the future.

Reply:
Yeah, that explicit fallback is being tracked in https://github.com/openshift/origin/pull/28000 which will eventually allow us to drop the fallback we currently have.